Bluetooth: Controller: Implement PAST Support

lutb-ot opened the pull request:
Implements:
  * LLCP PDU Flow and unittests for Periodic Sync Indication
  * Implements PAST support in ULL
  * PAST bsim tests
  * Fix PAST support bap_scan_delegator

aescolar commented on tests/bsim/bluetooth/ll/bis/testcase.yaml:
```diff
@@ -40,3 +40,16 @@ tests:
+    platform_allow:
+      - nrf52_bsim
+      - nrf5340bsim/nrf5340/
+      - nrf5340bsim/nrf5340/cpuapp
```
If you want to build for the nrf5340bsim/nrf5340/cpuapp you'll need to use sysbuild (you can check other bsim tests building for it)

aescolar commented on tests/bluetooth/controller/mock_ctrl/src/ull_sync.c:
```diff
@@ -0,0 +1,61 @@
+
+static struct ll_sync_set ll_sync[BT_PER_ADV_SYNC_MAX];
+
+/* TODO(lutb): remember to release sync after use - mem_release(sync, &sync_free); */
```
:)

lutb-ot replied:
Thanks, yes, that is a left over :) Good catch!

cvinayak reviewed (dismissed):
initial comments, will do re-review after these comments are addressed and CI jobs are passing.

cvinayak commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -48,6 +48,9 @@ config BT_LLL_VENDOR_NORDIC
 	select BT_CTLR_SCA_UPDATE_SUPPORT
+	select BT_CTLR_SYNC_TRANSFER_RECEIVER_SUPPORT
+	select BT_CTLR_SYNC_TRANSFER_SENDER_SUPPORT
+	select BT_TICKER_LAZY_GET
```
```suggestion
```

cvinayak replied:
There is line below like this `select BT_TICKER_LAZY_GET if BT_CTLR_ADV_PERIODIC || BT_CTLR_CENTRAL_ISO` ; append `|| BT_CTLR_SYNC_TRANSFER_SENDER`  to it

cvinayak reviewed (commented):
With a very quick pass, the changes LGTM, I do not see any objections from my side. Just get the CI happy :-)

cvinayak reviewed (commented):
nitpick.... use Bluetooth and Controller with `B` and `C`  in commit title and message body etc... (BT Spec documentation uses so).

lutb-ot commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4081,6 +4094,109 @@ static void le_read_pal_size(struct net_buf *buf, struct net_buf **evt)
+	uint16_t skip;
+	uint16_t timeout;
+	uint8_t  cte_type;
+	uint8_t status;
```
Noted, I'll make sure to get it fixed and follow that forward.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync_types.h:
```diff
@@ -94,15 +88,24 @@ struct ll_sync_set {
+	uint16_t conn_handle;
+	uint16_t service_data;
+};
+
+
```
```suggestion

```

Thalley commented on tests/bluetooth/controller/common/src/helper_pdu.c:
```diff
@@ -504,6 +504,33 @@ void helper_pdu_encode_sca_rsp(struct pdu_data *pdu, void *param)
+		sizeof(struct pdu_data_llctrl_periodic_sync_ind);
+	pdu->llctrl.opcode = PDU_DATA_LLCTRL_TYPE_PERIODIC_SYNC_IND;
+
+	pdu->llctrl.periodic_sync_ind.id = sys_cpu_to_le16(p->id);
+
+	memcpy(&pdu->llctrl.periodic_sync_ind.sync_info, &p->sync_info,
+			sizeof(struct pdu_adv_sync_info));
+
+	pdu->llctrl.periodic_sync_ind.conn_event_count = sys_cpu_to_le16(p->conn_event_count);
+	pdu->llctrl.periodic_sync_ind.last_pa_event_counter =
+							sys_cpu_to_le16(p->last_pa_event_counter);
+	pdu->llctrl.periodic_sync_ind.sid = p->sid;
+	pdu->llctrl.periodic_sync_ind.addr_type = p->addr_type;
+	pdu->llctrl.periodic_sync_ind.sca = p->sca;
+	pdu->llctrl.periodic_sync_ind.phy = p->phy;
+
+	memcpy(pdu->llctrl.periodic_sync_ind.adv_addr, p->adv_addr, sizeof(p->adv_addr));
+
+	pdu->llctrl.periodic_sync_ind.sync_conn_event_count = p->sync_conn_event_count;
```
Consider storing `pdu->llctrl.periodic_sync_ind` as a local variable to reduce line lengths

lutb-ot replied:
However, if this is changed it will not be consistent with the rest of the code in `helper_pdu.c`.

Thalley replied:
Welp :) 

It's OK to keep as is

Thalley commented on tests/bluetooth/controller/ctrl_periodic_sync/Kconfig:
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) 2022 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+# some of the control procedures in the BT LL depend on
+# the following configs been set
+
+config SOC_COMPATIBLE_NRF
+	default y
+
+config BT_CTLR_DATA_LEN_UPDATE_SUPPORT
+	default y
+
+config BT_CTLR_PHY_UPDATE_SUPPORT
+	default y
+
+config BT_CTLR_PHY_CODED_SUPPORT
+	default y
+
+config BT_CTLR_PHY_2M_SUPPORT
+	default y
+
+config ENTROPY_NRF_FORCE_ALT
+	default n
+
+config ENTROPY_NRF5_RNG
+	default n
```
What is preventing us from enabling these in `prj.conf`?

lutb-ot replied:
None, however this is the approach that is done for all the controller tests.

Thalley replied:
But I have to assume there were reasons why those couldn't be enabled in the other tests, otherwise this is also just a bug waiting to happen since we aren't respecting dependencies which can affect behavior of the functions we are testing. 

@cvinayak what's your though here?

cvinayak replied:
1. `_SUPPORT` ones are hidden and can only be overridden using an application Kconfig file.
2. the none hidden ones can be overridden in prj.conf, but these are again being `select` and hence had to be overridden by a Kconfig file, i am guessing.

Thalley commented on tests/bluetooth/controller/ctrl_periodic_sync/prj.conf:
```diff
@@ -0,0 +1,50 @@
+CONFIG_BT_CTLR_SYNC_TRANSFER_RECEIVER=y
+CONFIG_BT_CTLR_SYNC_ISO=y
+CONFIG_BT_CTLR_SCAN_SYNC_ISO_SET=1
+CONFIG_BT_CTLR_SYNC_ISO_STREAM_COUNT=1
+CONFIG_BT_CTLR_CONN_ISO_STREAMS=1
+CONFIG_BT_CTLR_CONN_ISO_GROUPS=1
```
Do we need ISO for this test?

lutb-ot replied:
Yes, as the `sca` field of the `struct lll_sync` is flaged with `#if defined(CONFIG_BT_CTLR_SYNC_ISO)` and is needed for PAST for work

Thalley replied:
Hmm, that sounds odd. ISO and PAST are unrelated features, and there shouldn't be a dependency there. Maybe the `sca` field need to be available for non-ISO builds? @cvinayak

cvinayak replied:
@lutb-ot you should make `sca`  available for PAST support too. 

Background, periodic sync needs `sca`  but uses it as local variable to calculate window widening. But a ISO sync is established using an inherited `sca` value, hence `sca` was stored in `lll_sync` context. Now that PAST feature will need the `sca` to be remembered, i think we should make the `sca` field available in the `lll_sync` context?  or wait, if this is for window widening, then it would be calculated only once, do we really need it to be present in `lll_sync` context?

Thalley replied:
Yeah, so `sca` should not depend on ISO, and thus should be updated any ISO requirements from these tests should be removed

lutb-ot replied:
I'll make `sca` depend on PAST too and update/fix ISO requirements in the bsim and unit tests.

lutb-ot replied:
Yes, I sadly didn't make it before my vacation as there where more dependencies that I thought. I have fixed it now.

Thalley commented on tests/bluetooth/controller/mock_ctrl/src/mayfly.c:
```diff
@@ -25,7 +25,8 @@ uint32_t mayfly_is_enabled(uint8_t caller_id, uint8_t callee_id)
 }
 
-uint32_t mayfly_enqueue(uint8_t caller_id, uint8_t callee_id, uint8_t chain, struct mayfly *m)
+__weak uint32_t mayfly_enqueue(uint8_t caller_id, uint8_t callee_id,
+							uint8_t chain, struct mayfly *m)
```
```suggestion
__weak uint32_t mayfly_enqueue(uint8_t caller_id, uint8_t callee_id,
			       uint8_t chain, struct mayfly *m)
```

Thalley commented on tests/bluetooth/controller/mock_ctrl/src/ull_adv_sync.c:
```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright (c) 2016 Nordic Semiconductor ASA
+ * Copyright (c) 2016 Vinayak Kariappa Chettimada
```
Should the copyrights be changed?

Thalley commented on tests/bsim/bluetooth/audio/src/bap_scan_delegator_test.c:
```diff
@@ -267,6 +268,14 @@ static int pa_sync_req_cb(struct bt_conn *conn,
+			err = bt_bap_scan_delegator_set_pa_state(state->recv_state->src_id,
+								 BT_BAP_PA_STATE_INFO_REQ);
+			if (err != 0) {
+				printk("Failed to set INFO_REQ state: %d",
+					    err);
```
```suggestion
				printk("Failed to set INFO_REQ state: %d", err);
```

Thalley commented on tests/bsim/bluetooth/audio/src/bap_scan_delegator_test.c:
```diff
@@ -267,6 +268,14 @@ static int pa_sync_req_cb(struct bt_conn *conn,
 
 	if (past_avail) {
 		err = pa_sync_past(conn, state, pa_interval);
+		if (err == 0) {
+			err = bt_bap_scan_delegator_set_pa_state(state->recv_state->src_id,
+								 BT_BAP_PA_STATE_INFO_REQ);
```
Very nice that you've added this support. Have you verified that it getting triggered correctly?

Thalley commented on tests/bsim/bluetooth/ll/bis/prj_past.conf:
```diff
@@ -0,0 +1,46 @@
+CONFIG_BT_EXT_ADV=y
+CONFIG_BT_PER_ADV=y
+CONFIG_BT_PER_ADV_SYNC=y
+CONFIG_BT_ISO_BROADCASTER=y
+CONFIG_BT_ISO_SYNC_RECEIVER=y
```
Is ISO needed for this test?

lutb-ot replied:
Yes, same as earlier message.

Thalley commented on tests/bsim/bluetooth/ll/bis/src/test_past.c:
```diff
@@ -0,0 +1,709 @@
+
+extern enum bst_result_t bst_result;
+
+#define FAIL(...)					\
+	do {						\
+		bst_result = Failed;			\
+		bs_trace_error_time_line(__VA_ARGS__);	\
+	} while (0)
+
+#define PASS(...)					\
+	do {						\
+		bst_result = Passed;			\
+		bs_trace_info_time(1, __VA_ARGS__);	\
+	} while (0)
```
Consider having a look at the https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/bsim/babblekit and https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/bluetooth/common/testlib to see if you can reduce the overhead here by using the predefined test helpers

lutb-ot replied:
This is how it is done everywhere, so it should probably be a separate PR for fixing and cleaning this up everywhere in both host and controller bsim tests.

Thalley replied:
I understand your point, although I don't necessarily agree. We do not need to follow the structure of older tests if new tests could use frameworks that were not available when the older tests were written. 

It's OK to keep as is, but could have saved you time to use the frameworks we have now :)

Thalley reviewed (changes requested):
Very nice work. Great that you've added a bunch of tests for this

Most comments are code style and security related

lutb-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -2545,6 +2604,151 @@ void ull_conn_default_tx_time_set(uint16_t tx_time)
+		ticker_offset_us = get_ticker_offset(TICKER_ID_SCAN_SYNC_BASE + sync_handle,
+						     &lazy);
+
+		if (lazy && ticker_offset_us > interval_us) {
+
+			/* Figure out how many events we have actually skipped */
+			lazy = lazy - (ticker_offset_us / interval_us);
```
Yes, `lazy` is number of skip events

So in the case that we are actually skipping events, we need to know how many events we have skipped so far to determine the correct `pa_event_counter` and `last_pa_event_counter.`

Thalley replied:
Can we get a better name for the variable? `lazy` is a adjective which is usually only for `bool` values. The variable name should indicate that it's a number of events

lutb-ot replied:
`lazy` is used through out the controller to describe the latency/skip events, so it was more to keep it consistent.

Thalley replied:
Fair, that makes sense. Consistency is good :) @cvinayak would it make sense to do a controller-wide rename here?

cvinayak replied:
`lazy` term used in the Controller's scheduling is an idealogy similar to https://en.wikipedia.org/wiki/Lazy_evaluation but for control flow. It is more than latency or skipped of events; it also represents a weight/rank in the scheduling strategy, overlap/collision count and elapsed intervals (i.e. between dissimilar ACL and ISO interval). `lazy_count` or `lazy_value` could be better names.

But this outside the scope of this PR to make a Controller-wide rename, and should be a different PR.

Thalley replied:
Agreed. As long we can move away from an adjective for a count value I'm happy :)

lutb-ot commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_pdu.h:
```diff
@@ -195,6 +195,11 @@ lll_adv_sync_data_latest_peek(const struct lll_adv_sync *const lll)
 	return lll_adv_pdu_latest_peek(&lll->data);
 }
 
+static inline struct pdu_adv *lll_adv_sync_data_curr_get(struct lll_adv_sync *lll)
+{
+	return (void *)lll->data.pdu[lll->data.first];
+}
```
@cvinayak  will you be able to add more information here, as this function follows the same design as `struct pdu_adv *lll_adv_aux_data_curr_get(struct lll_adv_aux *lll)`

cvinayak replied:
The ADV PDU are double buffered. `first`  buffer remains persistent without being modified until the LLL execution consumes it, `first` is modified by LLL execution context and safely readable by both ULL and LLL. 

`last` is modified by ULL context and safely readable by both ULL and LLL.

LLL always peeks the latest in `last` index... i.e. it modifies `first` until it is `last`  index when dequeueing, then uses the data in the `last` index.

The `_curr_get()` is the naming used for the first data placed by ULL context that is not yet consumed by LLL. In the context of PAST only the AdvA is read back, which will anyway be same in case application is continuously updating data before LLL dequeues it, hence the name and its use is correct here.

Thalley replied:
Is that documented anywhere? Just for future references

cvinayak replied:
I thought we had something, but seems it is not upstreamed, hence create this: https://github.com/zephyrproject-rtos/zephyr/issues/79643

lutb-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -967,6 +1023,7 @@ void ull_adv_sync_pdu_init(struct pdu_adv *pdu, uint8_t ext_hdr_flags,
 #endif /* CONFIG_BT_CTLR_ADV_PERIODIC_ADI_SUPPORT */
 				     ULL_ADV_PDU_HDR_FIELD_SYNC_INFO)));
 
+#if defined(CONFIG_BT_CTLR_ADV_SYNC_PDU_BACK2BACK)
```
Using IS_ENABLED here will not be feasible here, as `cte_len_us` and `aux_ptr` wouldn't be defined leading to build errors.

lutb-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -1607,6 +1658,14 @@ static int init_reset(void)
 #endif /* CONFIG_BT_CTLR_PHY_CODED */
 #endif /* CONFIG_BT_CTLR_PHY */
 
+#if defined(CONFIG_BT_CTLR_SYNC_TRANSFER_RECEIVER)
+	default_mode     = 0;
+	/* Default 6 events */
+	default_skip     = 6;
+	default_timeout  = 10;
+	default_cte_type = 0;
```
They are chosen by me and no default values are defined in the core spec.

Thalley replied:
> They are chosen by me as no default values are defined in the core spec.

Would it make sense to make this Kconfig configurable? 

The spec does not suggest any initial values for PAST if this is not set explicitly by the host?

lutb-ot replied:
The default values here are just dummy values as `default_mode` set to 0 means the PAST is disabled `BT_HCI_LE_PAST_MODE_NO_SYNC.` Hence, the host will need to subscribe to PAST using `bt_le_per_adv_sync_transfer_subscribe` in order for PAST to be enabled by changing the `default_mode` and thereby also the `default_skip`, `default_timeout` and `default_cte_type`.

It might therefore be better to set everything to zero instead.

Thalley replied:
> It might therefore be better to set everything to zero instead.

If `bt_le_per_adv_sync_transfer_subscribe` is required to be called by the host, then there's no value to have default values here (and may be harmful as it indicates that _some_ values are set). I suggest to leave them all as `0`.

lutb-ot replied:
Agree, done

lutb-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -2545,6 +2604,151 @@ void ull_conn_default_tx_time_set(uint16_t tx_time)
+				       ticker_get_offset_op_cb, (void *)&ret_cb);
+
+	if (ret == TICKER_STATUS_BUSY) {
+		while (ret_cb == TICKER_STATUS_BUSY) {
+			ticker_job_sched(TICKER_INSTANCE_ID_CTLR, TICKER_USER_ID_ULL_LOW);
+		}
```
This approach is done several places in the controller and to get the most accurate and correct answer @cvinayak is probably the right one to answer this. :)

cvinayak replied:
This is a busy wait implementation for an interrupt (software interrupt) being triggered but to be enabled by another interrupt (radio interrupt). This is in case of nrf52 never entering the if-clause as the software interrupt is never disabled. This could only be a forever loop for nrf51, but that the case where a radio interrupt never arrives in which case its a h/w issue anyway.

lutb-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_past.c:
```diff
@@ -0,0 +1,570 @@
+				ctx->state = RP_PAST_STATE_WAIT_RESOLVE_INTERFACE_AVAIL;
+				return;
+			}
+#else
+			/* TODO: Not implemented - use RPA */
+#endif /* CONFIG_BT_CTLR_SW_DEFERRED_PRIVACY */
```
It will still work, but will not be 100% spec compliant. Therefore an assert might be overkill. :)

lutb-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -2545,6 +2604,151 @@ void ull_conn_default_tx_time_set(uint16_t tx_time)
+				      last_pa_event_counter);
+}
+
+void ull_conn_past_sender_offset_get(struct ll_conn *conn)
```
I understand how this naming can be confusing, however it does a `mayfly_enqueue`. When it runs it will send the result back using `ull_lp_past_offset_calc_reply`, which will set the result into the PDU and change the state in the llcp PAST state machine.

Would `ull_conn_past_sender_offset_request` be better?

lutb-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -1322,6 +1444,92 @@ bool ull_lp_cc_is_enqueued(struct ll_conn *conn)
+
+		ctx->data.periodic_sync.sync_info.evt_cntr = sys_cpu_to_le16(pa_event_counter);
+
+		ctx->data.periodic_sync.last_pa_event_counter = last_pa_event_counter;
```
They all need to be stored as LE. However, the other ones, namely `last_pa_event_counter`, `conn_event_count` and  `sync_conn_event_count` are converted in `llcp_pdu_encode_periodic_sync_ind()` and then decoded on the receiving side in `llcp_pdu_decode_periodic_sync_ind`. The reason for `evt_cntr` being converted here is that sync_info is `memcpy` in `llcp_pdu_encode_periodic_sync_ind()` and therefore needs to be converted here.

Thalley replied:
Hmm, not a fan of this. If they all need to be stored as LE, then why not do it here rather than in `llcp_pdu_decode_periodic_sync_ind`? 

Or rather, if the others are converted to LE in `llcp_pdu_decode_periodic_sync_ind`, why not just convert `evt_cntr` there too?

lutb-ot replied:
I agree, lets go with converting all in `llcp_pdu_encode_periodic_sync_ind`

lutb-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_past.c:
```diff
@@ -0,0 +1,570 @@
+
+		if (ctx->data.periodic_sync.addr_type == BT_ADDR_LE_RANDOM) {
+			/* TODO: For efficiency, check if address resolve is needed */
+#if defined(CONFIG_BT_CTLR_SW_DEFERRED_PRIVACY)
```
It is sadly not feasible as `rp_past_addr_resolve()` is `#ifdef`, hence it will lead to compilation errors.

lutb-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_past.c:
```diff
@@ -0,0 +1,570 @@
+#endif /* CONFIG_BT_CTLR_SW_DEFERRED_PRIVACY */
+		}
+
+#if defined(CONFIG_BT_PERIPHERAL)
+		/* If sync_conn_event_count is this connection event,
+		 * we have to wait for drift correction for this event to be applied -
+		 * continue processing in the next conn event
+		 */
+		if (conn->lll.role == BT_HCI_ROLE_PERIPHERAL &&
+		   ctx->data.periodic_sync.sync_conn_event_count == ull_conn_event_counter(conn)) {
+			ctx->state = RP_PAST_STATE_WAIT_NEXT_EVT;
+			return;
+		}
+#endif /* CONFIG_BT_PERIPHERAL */
```
Yes, that is feasible.

lutb-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:
```diff
@@ -190,7 +190,7 @@ static void pu_reset_timing_restrict(struct ll_conn *conn)
 
 #if defined(CONFIG_BT_PERIPHERAL)
-static inline bool phy_valid(uint8_t phy)
+static inline bool phy_validation(uint8_t phy)
```
It is a different PHY validation, see LL_PHY_CHECK_IND: 

_"If the Peripheral receives an LL_PHY_UPDATE_IND where either PHY field specifies a PHY that the Peripheral does not support, has a bit set that is reserved for future use, or has more than one bit set, the Peripheral shall not change the PHY in that direction."_

For LL_PERIODIC_SYNC_IND is: 
_"However, if the PHY field of the LL_PERIODIC_SYNC_IND PDU or the LL_PERIODIC_SYNC_WR_IND PDU has no bits or more than one bit set, or the bit set corresponds to a PHY that the recipient does not support or is reserved for future use, the recipient shall ignore the PDU"_

Thalley replied:
Aha. 

We should not have a function called `phy_valid` that works on one set of PHY values, and a function `phy_validation` that works on another set of PHY values. 

The function names should indicate the behavior of them, and since they both have the same name and different behavior, that's just a bug waiting to happen

lutb-ot replied:
The `phy_validation` is static and can only be used in ull_llcp_phy.c 

Where `phy_valid` is shared for several procedures as it is a more general check. However, I get that the name can give problmes. I can try to get up with a better name for the ` ull_llcp_phy.c ` specific version.

Thalley replied:
Thanks :)

lutb-ot commented on subsys/bluetooth/controller/ll_sw/ull_sync.c:
```diff
@@ -884,24 +1160,21 @@ void ull_sync_setup(struct ll_scan_set *scan, struct ll_scan_aux_set *aux,
+	struct ll_scan_set *scan;
+
 	/* Remove the sync context from being associated with scan contexts */
+	scan = ull_scan_set_get(SCAN_HANDLE_1M);
```
Yes, and we also check for the coded phy.

lutb-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -2545,6 +2604,151 @@ void ull_conn_default_tx_time_set(uint16_t tx_time)
+		ticker_offset_us = get_ticker_offset(TICKER_ID_ADV_SYNC_BASE + adv_sync_handle,
+						     &lazy);
+
+		pa_event_counter = adv_sync->lll.event_counter;
+		last_pa_event_counter = pa_event_counter - 1;
```
Yes, in order to send a PAST a sync will have to be established, hence `pa_event_counter` > 0

lutb-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -2545,6 +2604,151 @@ void ull_conn_default_tx_time_set(uint16_t tx_time)
+		/* Correct for window widening */
+		ticker_offset_us += window_widening_event_us;
+
+		pa_event_counter = sync->lll.event_counter + lazy;
+
+		last_pa_event_counter = pa_event_counter - 1 - lazy;
```
Same as above

lutb-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -2545,6 +2604,151 @@ void ull_conn_default_tx_time_set(uint16_t tx_time)
+	hal_ticker_add_jitter(&ticks_to_expire, &remainder);
+	start_us = remainder;
+
+	return HAL_TICKER_TICKS_TO_US(ticks_to_expire) + remainder_us - start_us;
```
Yes, good catch. Will add fix using `isoal_get_wrapped_time_us`

lutb-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -1322,6 +1444,92 @@ bool ull_lp_cc_is_enqueued(struct ll_conn *conn)
+		if (conn->lll.role == BT_HCI_ROLE_PERIPHERAL && done->extra.trx_cnt) {
+			uint32_t start_to_actual_us;
+
+			start_to_actual_us = done->extra.drift.start_to_address_actual_us;
+			start_to_actual_us -= done->extra.drift.preamble_to_addr_us;
+
```
Will add check - `isoal_get_wrapped_time_us`

lutb-ot commented on subsys/bluetooth/controller/ll_sw/ull_sync.c:
```diff
@@ -279,6 +200,245 @@ uint8_t ll_sync_create(uint8_t options, uint8_t sid, uint8_t adv_addr_type,
+
+	sync_offset_us = PDU_ADV_SYNC_INFO_OFFSET_GET(si) * lll->window_size_event_us;
+	/* offs_adjust may be 1 only if sync setup by LL_PERIODIC_SYNC_IND */
+	sync_offset_us += (PDU_ADV_SYNC_INFO_OFFS_ADJUST_GET(si) ? OFFS_ADJUST_US : 0U);
+	sync_offset_us -= EVENT_TICKER_RES_MARGIN_US;
+	sync_offset_us -= EVENT_JITTER_US;
+	sync_offset_us -= ready_delay_us;
```
No risk here, as `PDU_ADV_SYNC_INFO_OFFSET_GET(si) * lll->window_size_event_us ` is minimum 300us, hence `EVENT_TICKER_RES_MARGIN_US`, `EVENT_JITTER_US` and `ready_delay_us` is below that.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync.c:
```diff
@@ -27,30 +27,48 @@
 #include <soc.h>
 #include "hal/debug.h"
 
+#define ABS(x) ((x) > 0 ? (x) : -(x))
```
We do actually have `abs` as part of the `stdlib.h`. 

I suggest to use that instead of re-implementing that function as a macro here

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_types.h:
```diff
@@ -4,6 +4,8 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+ #define INVALID_ADV_HANDLE 0xFF
+
```
We already have `BT_HCI_ADV_HANDLE_INVALID` which we can use

lutb-ot replied:
My bad, I'll fix that.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync_types.h:
```diff
@@ -8,6 +8,8 @@
 #define LL_SYNC_STATE_ADDR_MATCH 0x01
 #define LL_SYNC_STATE_CREATED    0x02
 
+ #define INVALID_SYNC_HANDLE     0xFFFF
+
```
We have `BT_HCI_SYNC_HANDLE_INVALID` that you can use instead

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn_types.h:
```diff
@@ -158,6 +158,15 @@ struct ll_conn {
 	struct lll_conn lll;
 
+#if defined(CONFIG_BT_CTLR_SYNC_TRANSFER_RECEIVER)
+	struct {
+		uint8_t mode;
+		uint8_t cte_type;
+		uint16_t skip;
+		uint16_t timeout;
+	} past;
```
Would it make sense to define this struct by itself and use that in `ull_conn`, instead of the 4 individual values? Would make it easier to reset, and could even return those values as a reference or a single value instead of having 4 functions so that

```c
	conn->past.mode    = ull_conn_default_past_param_mode_get();
	conn->past.skip    = ull_conn_default_past_param_skip_get();
	conn->past.timeout = ull_conn_default_past_param_timeout_get();
	conn->past.cte_type = ull_conn_default_past_param_ctetype_get();
```

could become something like 

```c
	conn->past = ull_conn_default_past_param_get();
```
or

```c
	memcpy(&conn->past, ull_conn_default_past_param_get(), sizeof(conn->past));
```

or is there some value in having 4 separate functions and values for this?

lutb-ot replied:
No, I think it will make perfect sense to make it into a single get function instead of four.

I'll make a change for that :)

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -1019,6 +1034,114 @@ uint8_t ull_cp_conn_update(struct ll_conn *conn, uint16_t interval_min, uint16_t
+							INVALID_SYNC_HANDLE;
+	ctx->data.periodic_sync.adv_handle = adv_sync ? ull_adv_sync_handle_get(adv_sync) :
+							INVALID_ADV_HANDLE;
+	ctx->data.periodic_sync.id = service_data;
```
Would it make sense to use `service_data` instead of `id` in the `periodic_sync` struct?

lutb-ot replied:
The only reason it is called ID is because of the naming in the Spec:

"ID shall be set to an identifier provided by the Host. This is for use by higher-level protocols and its value is not specified or used by this specification."

Thalley replied:
Aha, interesting. I'm having trouble finding where it's defined that the `id` here is the same value as the `service_data` provided by the host. I believe it's correct that it is, but haven't found where that's explicitly stated

lutb-ot replied:
I haven't seen it explicitly stated anywhere in the Spec. However, it is the only host parameter in `HCI_LE_­Periodic_­Advertising_­Sync_­Transfer` and `HCI_LE_­Periodic_­Advertising_­Sync_­Transfer` with the correct range and description, but it is not explicitly stated anywhere.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -1840,3 +2060,11 @@ struct proc_ctx *llcp_create_procedure(enum llcp_proc proc)
+	/* This is equivalent to:
+	 * exactly one bit set, and no bit set is rfu's
+	 */
+	return (phy == PHY_1M || phy == PHY_2M || phy == PHY_CODED);
```
Should `PHY_2M` be considered valid if `CONFIG_BT_CTLR_PHY_2M=n`? Ditto with coded?

lutb-ot replied:
It will be checked after: see: `ull_llcp_past.c` line:83:

`#if defined(CONFIG_BT_CTLR_PHY)
	const uint8_t phy = pdu->llctrl.periodic_sync_ind.phy;

	if (((phy & PHY_2M) && !IS_ENABLED(CONFIG_BT_CTLR_PHY_2M)) ||
	    ((phy & PHY_CODED) && !IS_ENABLED(CONFIG_BT_CTLR_PHY_CODED))) {
		/* Unsupported phy selected */
		return BT_HCI_ERR_UNSUPP_FEATURE_PARAM_VAL;
	}
#endif /* CONFIG_BT_CTLR_PHY */`

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync.c:
```diff
@@ -279,6 +199,245 @@ uint8_t ll_sync_create(uint8_t options, uint8_t sid, uint8_t adv_addr_type,
+	conn_interval_us = conn->lll.interval * CONN_INT_UNIT_US;
+
+	/* Calculate offset and schedule sync radio events */
+	ready_delay_us = lll_radio_rx_ready_delay_get(lll->phy, 1);
```
```suggestion
	ready_delay_us = lll_radio_rx_ready_delay_get(lll->phy, PHY_FLAGS_S8);
```

Right?

lutb-ot replied:
Yes, that should be the most correct way. Will change that.

cvinayak replied:
yes, please update (atleast in new added code).

lutb-ot committed:
Bluetooth: Controller: Implement Periodic Sync Procedure

Adding PDU flow for Periodic Sync Indication

Signed-off-by: Lucas Mathias Balling <lutb@demant.com>

lutb-ot committed:
tests: Bluetooth: LLCP unittests for Periodic Sync Procedure

Added LLCP PDU Periodic Sync Indication unittests

Signed-off-by: Lucas Mathias Balling <lutb@demant.com>

lutb-ot committed:
Bluetooth: Controller: Implement PAST support in ULL

Implement PAST support in ULL and fixed test after changed
dependencies in ull_sync_internal.c

Signed-off-by: Lucas Mathias Balling <lutb@demant.com>

lutb-ot committed:
tests: Bluetooth: bsim tests for Periodic Sync Transfer

Implemented:
  * Central sync to broadcaster, connects to peripheral and
    transfers sync. Peripheral Syncs to PA
  * Broadcaster acting Central connects to peripheral
    and transfers sync. Peripheral sync to PA
  * Moved previous bis tests to test_bis.c and
    new PAST tests was added in test_past.c

Signed-off-by: Lucas Mathias Balling <lutb@demant.com>

lutb-ot committed:
Bluetooth: Audio: Fix PAST support for bap_scan_delegator

Fixed PAST support for bap_scan_delegator

Signed-off-by: Lucas Mathias Balling <lutb@demant.com>

Thalley reviewed (commented):
No more blocking comments from me, but we should really get @cvinayak to do another review and respond to some of the outstanding comments before I'll approve

cvinayak reviewed (dismissed):
No definite change requests from my end. LGTM.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_sync.c:
```diff
@@ -279,6 +199,243 @@ uint8_t ll_sync_create(uint8_t options, uint8_t sid, uint8_t adv_addr_type,
+	       PDU_SYNC_INFO_SCA_CHM_SCA_BIT_MASK) >>
+	      PDU_SYNC_INFO_SCA_CHM_SCA_BIT_POS;
+
+	lll->sca = sca;
```
dont you need conditional compiles here?

lutb-ot replied:
The whole function is already flaged with CONFIG_BT_CTLR_SYNC_TRANSFER_RECEIVER.

cvinayak reviewed (dismissed):
LGTM, just need to get the CI happy...

Thalley commented:
@lutb-ot I don't want you to update and dismiss the approvals here, but this feature really should be mentioned in the release notes :) Please follow up with a PR to add it there after this has been merged
